src/dos/drive_fat.cpp

```diff
@@ -1056,15 +1082,24 @@ bool fatDrive::FileExists(const char *name) {
 bool fatDrive::FileOpen(DOS_File **file, char *name, uint32_t flags) {
 	direntry fileEntry;
 	uint32_t dirClust, subEntry;
-	if(!getFileDirEntry(name, &fileEntry, &dirClust, &subEntry)) return false;
-	/* TODO: check for read-only flag and requested write access */
+
+	if (!getFileDirEntry(name, &fileEntry, &dirClust, &subEntry)) {
+		DOS_SetError(DOSERR_FILE_NOT_FOUND);
+		return false;
+	}
+
+	if ((readonly || (fileEntry.attrib & DOS_ATTR_READ_ONLY)) &&
+	    (flags & 0xf) != OPEN_READ) {
+		DOS_SetError(DOSERR_ACCESS_DENIED);
+		return false;
+	}
+
 	*file = new fatFile(name, fileEntry.loFirstClust, fileEntry.entrysize, this);
 	(*file)->flags = flags;
 	((fatFile *)(*file))->dirCluster = dirClust;
 	((fatFile *)(*file))->dirIndex = subEntry;
-	/* Maybe modTime and date should be used ? (crt matches findnext) */
-	((fatFile *)(*file))->time = fileEntry.crtTime;
-	((fatFile *)(*file))->date = fileEntry.crtDate;
+	((fatFile *)(*file))->time = fileEntry.modTime;
```

Comment:
After the flags are set, suggest a single C++ reinterpret cast over to a `fat_file` pointer and then use that in all four subsequent `((fatFile *)(*file))->` calls.